enclave/sgx/collateralinfo.c

```diff
@@ -169,6 +170,123 @@ oe_result_t oe_get_sgx_quote_verification_collateral(
+    size_t prealloc_qe_identity_issuer_chain_size = 3000;
+    size_t prealloc_qe_identity_size = 700;
+    if(!buf->tcb_info_size)
+    {
+        buf->tcb_info = (uint8_t *)malloc(prealloc_tcb_info_size);
+        if(buf->tcb_info != NULL)
+        {
+            buf->tcb_info_size = prealloc_tcb_info_size;
+            memset(buf->tcb_info,0, buf->tcb_info_size);   
+        }
+        else
+        {
+            goto cleanup;
+        }        
+    }
+
+    if(!buf->tcb_info_issuer_chain_size)
+    {
+        buf->tcb_info_issuer_chain = (uint8_t *)malloc(prealloc_tcb_info_size);
+        if(buf->tcb_info != NULL)
+        {
+            buf->tcb_info_issuer_chain_size = prealloc_tcb_info_issuer_chain_size;
+            memset(buf->tcb_info_issuer_chain,0, buf->tcb_info_issuer_chain_size);    
+        }
+        else
+        {
+            goto cleanup;
+        }
+    }
+
+    if(!buf->pck_crl_size)
+    {
+        buf->pck_crl = (uint8_t *)malloc(prealloc_pck_crl_size);
+        if(buf->pck_crl != NULL)
+        {
+            buf->pck_crl_size = prealloc_pck_crl_size;
+            memset(buf->pck_crl,0, buf->pck_crl_size);
+        }
+        else
+        {
+            goto cleanup;
+        }
+    }
+
+    if(!buf->root_ca_crl_size)
+    {
+        buf->root_ca_crl = (uint8_t *)malloc(prealloc_root_ca_crl_size);
+        if(buf->root_ca_crl != NULL)
+        {
+            buf->root_ca_crl_size = prealloc_root_ca_crl_size;
+            memset(buf->root_ca_crl,0, buf->root_ca_crl_size);
+        }
+        else
+        {
+            goto cleanup;
+        }
+    }
+
+    if(!buf->pck_crl_issuer_chain_size)
+    {
+        buf->pck_crl_issuer_chain = (uint8_t *)malloc(prealloc_pck_crl_issuer_chain_size);
+        if(buf->pck_crl_issuer_chain != NULL)
+        {
+            buf->pck_crl_issuer_chain_size = 
+            prealloc_pck_crl_issuer_chain_size;
+            memset(buf->pck_crl_issuer_chain,0, buf->pck_crl_issuer_chain_size
+            );
+        }
+        else
+        {
+            goto cleanup;
+        }
+    }
+
+    if(!buf->qe_identity_size)
+    {   
+        buf->qe_identity = (uint8_t *)malloc(prealloc_qe_identity_size);
+        if(buf->qe_identity != NULL)
+        {
+            buf->qe_identity_size = prealloc_qe_identity_size;
+            memset(buf->qe_identity,0, buf->qe_identity_size);
+        }
+        else
+        {
+            goto cleanup;
+        }
+    }   
+
+    if(!buf->qe_identity_issuer_chain_size)
+    {
+        buf->qe_identity_issuer_chain = (uint8_t *)malloc(prealloc_qe_identity_issuer_chain_size);
+        if(buf->qe_identity_issuer_chain != NULL)
+        {
+            buf->qe_identity_issuer_chain_size = prealloc_qe_identity_issuer_chain_size;
+            memset(buf->qe_identity_issuer_chain,0, buf->qe_identity_issuer_chain_size);
+        }
+        else
+        {
+            goto cleanup;
+        }
+    }
+    return;
```

Comment:
Avoid early return when using labels/gotos. goto is generally scary so when it is being used it should be the only point of exit from the function.

`free` checks for null pointers anyway so it's safe to call free on a null pointer, this is ensured by POSIX.

Reply:
I had to use early return because the goto set of code is only supposed to execute for unsuccessful allocations. I agree about avoiding using goto statements all together. I can use a function if you suggest. I was trying to keep it short.